### DISCUSSION TITLE
Combination of proxyWithHistory and subscribe

### DISCUSSION THREAD

**Discussion by LukasSchwarz4202** [pmndrs/valtio#506] [Q&A] [answered]
Hello,

first of all thanks for all the great work done her with valtio!

I am currently changing part of the code of an existing application from using `proxy` to `proxyWithHistory`.
Doing so I was hoping that it would work together with the `subscribe` mechanism. In detail when `undo()` or `redo()` changes an element a component is subscribed to, that this subscribe function would be called. Apparently it is not.

I recreated the problem with codesandbox.io: https://codesandbox.io/s/valtio-subscribe-proxywithhistory-eir1bo

- Pressing **ChangeKeyTwoValue1AndSaveHistroy** changes an element and calles saveHistory() => subscribe() is called as expected.
- Pressing **ChangeKeyTwoValue2AndSaveHistroy** changes an element and calles saveHistory() => subscribe() is called as expected.
- Pressing **Undo** does not call subscribe()

As this is a post in the Q&A now my question: are my expectations wrong and as soon as `undo()` or `redo()` is called do I have to `unsubscribe()` and `subscribe()` again.

Thanks a lot in advance
Lukas

**Comment by dai-shi** [comment, marked as answer]
`undo()` will replace `keyValueProxy.value` with a new object, so you have to subscribe to `keyValueProxy` to get notified with undo.
https://codesandbox.io/s/valtio-subscribe-proxywithhistory-forked-uk394m
(If you need to track some specific properties, you would need to keep the previous value and compare.)

**Reply by LukasSchwarz4202** [reply]
Thanks for the fast reply
